### PR TITLE
Payments

### PR DESCRIPTION
**Pull request by hei-ryan**

### CODE REVIEW

File: doc/api.yml [hunk 1]
```diff
@@ -658,6 +658,15 @@ paths:
           schema:
             type: string
       operationId: createStudentFees
+      requestBody:
+        description: Student fees to create
```
Comment (hei-teacher) [inline doc/api.yml]: No.

**Review by hei-teacher — CHANGES_REQUESTED** [review]

File: doc/api.yml [hunk 2]
```diff
@@ -738,7 +756,38 @@ paths:
...
+          $ref: '#/components/responses/400'
+        '403':
+          $ref: '#/components/responses/403'
+        '404':
+          $ref: '#/components/responses/404'
+        '429':
+          $ref: '#/components/responses/429'
+        '500':
+          $ref: '#/components/responses/500'
+  /students/{student_id}/fees/payments:
```
Comment (hei-teacher) [inline doc/api.yml]: No! `fees/{fee_id}/payments`

File: src/test/java/school/hei/haapi/integration/PaymentIT.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        .type(Payment.TypeEnum.CASH)
+        .amount(2000)
+        .comment("Comment")
+        .creationDatetime(Instant.parse("2021-11-08T08:25:24.00Z"));
+  }
+
+  static Payment payment2() {
+    return new Payment()
+        .id("payment2_id")
+        .feeId("fee2_id")
```
Comment (hei-teacher) [inline src/test/java/school/hei/haapi/integration/PaymentIT.java]: Also implement basic connection between fee and payment. And corresponding IT.

**Review by hei-teacher — CHANGES_REQUESTED** [review]

File: src/test/java/school/hei/haapi/integration/PaymentIT.java (hunk 3 above)
Comment (hei-ryan) [reply src/test/java/school/hei/haapi/integration/PaymentIT.java]: Basic connection ?

File: doc/api.yml [hunk 4]
```diff
@@ -686,12 +694,50 @@ paths:
...
-
+  /fees:
+    get:
+      tags:
+        - Paying
+      summary: Get all student fees filtered by status
+      parameters:
+        - name: status
+          in: query
+          description: "Filter fees by status. Available status are PAID, UNPAID and LATE"
```
Comment (hei-teacher) [inline doc/api.yml:705]: Do not enumerate statuses here as they may change. Just say: `See the PaymentStatus object for its value.`

File: doc/api.yml [hunk 5]
```diff
@@ -686,12 +694,50 @@ paths:
...
+          schema:
+            $ref: '#/components/schemas/Page'
+        - name: page_size
+          in: query
+          schema:
+            $ref: '#/components/schemas/PageSize'
+      operationId: getFees
+      responses:
+        '200':
+          description: List of all student fees
```
Comment (hei-teacher) [inline doc/api.yml:719]: of filtered student fees

File: doc/api.yml [hunk 6]
```diff
@@ -686,12 +694,50 @@ paths:
...
+        '429':
+          $ref: '#/components/responses/429'
+        '500':
+          $ref: '#/components/responses/500'
   /students/{student_id}/fees/{fee_id}/payments:
     get:
       tags:
         - Paying
-      summary: Get all student payments
+      summary: Get all student payments for one specific fee ordered by creation datetime desc
```
Comment (hei-teacher) [inline doc/api.yml:740]: of as specific fee. Resulting list is ordered by...

File: doc/api.yml [hunk 7]
```diff
@@ -703,7 +749,15 @@ paths:
...
-      operationId: getStudentPayments
+        - name: page
+          in: query
+          schema:
+            $ref: '#/components/schemas/Page'
+        - name: page_size
+          in: query
+          schema:
+            $ref: '#/components/schemas/PageSize'
+      operationId: getStudentFeePayments
```
Comment (hei-teacher) [inline doc/api.yml:760]: getStudentPayments

File: doc/api.yml [hunk 8]
```diff
@@ -739,15 +793,24 @@ paths:
...
+        content:
+          application/json:
+            schema:
+              type: array
+              items:
+                $ref: '#/components/schemas/CreatePayment'
       responses:
         '200':
-          description: Created student payments
+          description: Created student payments,
```
Comment (hei-teacher) [inline doc/api.yml:807]: rm `,`

File: src/main/java/school/hei/haapi/endpoint/rest/controller/PaymentController.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+      @RequestBody List<CreatePayment> toCreate) {
+    return paymentService
+        .saveAll(paymentMapper.toDomainPayment(feeId, toCreate))
+        .stream()
+        .map(paymentMapper::toRestPayment)
+        .collect(toUnmodifiableList());
+  }
+
+  @GetMapping("/students/{studentId}/fees/{feeId}/payments")
+  public List<Payment> getFeePaymentsByStudentId(
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/endpoint/rest/controller/PaymentController.java:39]: getPaymentsByStudentId

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeMapper.java [hunk 10]
```diff
@@ -1,32 +1,44 @@
...
+import school.hei.haapi.service.UserService;
 
 @Component
+@AllArgsConstructor
 public class FeeMapper {
+
+  private final UserService userService;
+
   public Fee toRestFee(school.hei.haapi.model.Fee fee) {
+    Instant truncatedInstant = fee.getCreationDatetime().truncatedTo(ChronoUnit.MILLIS);
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeMapper.java:23]: No. I think we should of a more global way to handle this issue.

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeMapper.java [hunk 11]
```diff
@@ -1,32 +1,44 @@
...
         .totalAmount(fee.getTotalAmount())
         .remainingAmount(fee.getRemainingAmount())
         .comment(fee.getComment())
-        .creationDatetime(fee.getCreationDatetime())
+        .creationDatetime(truncatedInstant)
         .dueDatetime(fee.getDueDatetime());
   }
 
-  public school.hei.haapi.model.Fee toDomainFee(CreateFee createFee) {
+  private school.hei.haapi.model.Fee toDomainFee(User student, CreateFee createFee) {
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeMapper.java:36]: Be defensive. Throw a BadRequest if student.role is not STUDENT

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeMapper.java [hunk 12]
```diff
@@ -35,6 +47,18 @@ public school.hei.haapi.model.Fee toDomainFee(CreateFee createFee) {
         .build();
   }
 
+  public List<school.hei.haapi.model.Fee> toDomainFee(String studentId, List<CreateFee> toCreate) {
+    User student = userService.getById(studentId);
+    if (student == null) {
+      throw new BadRequestException("Student.id=[" + studentId + "] is not found");
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeMapper.java:53]: NotFoundException. Which takes the resource type as argument if I remember correctly.

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeMapper.java [hunk 13]
```diff
@@ -35,6 +47,18 @@ public school.hei.haapi.model.Fee toDomainFee(CreateFee createFee) {
...
+    }
+    List<school.hei.haapi.model.Fee> fees = new ArrayList<>();
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeMapper.java:55]: Use functional style. stream + map

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/PaymentMapper.java [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+import school.hei.haapi.model.exception.BadRequestException;
+import school.hei.haapi.service.FeeService;
+
+@Component
+@AllArgsConstructor
+public class PaymentMapper {
+  private final FeeService feeService;
+
+  public Payment toRestPayment(school.hei.haapi.model.Payment payment) {
+    Instant truncatedInstant = payment.getCreationDatetime().truncatedTo(ChronoUnit.MILLIS);
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/endpoint/rest/mapper/PaymentMapper.java:21]: No.

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/PaymentMapper.java [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+        .type(payment.getType())
+        .amount(payment.getAmount())
+        .comment(payment.getComment())
+        .creationDatetime(truncatedInstant);
+  }
+
+  private school.hei.haapi.model.Payment toDomainPayment(
+      Fee associatedFee, CreatePayment createPayment) {
+    if (createPayment.getAmount() == null) {
+      throw new BadRequestException("Amount is mandatory");
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/endpoint/rest/mapper/PaymentMapper.java:34]: Not here. Create a CreatePayment Validator

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/PaymentMapper.java [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+    }
+    List<school.hei.haapi.model.Payment> payments = new ArrayList<>();
+    for (CreatePayment c : createPayment) {
+      payments.add(toDomainPayment(associatedFee, c));
+    }
+    return payments;
+  }
+
+  private Payment.TypeEnum toDomainPaymentType(CreatePayment.TypeEnum createPaymentType) {
+    String paymentType = createPaymentType.toString();
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/endpoint/rest/mapper/PaymentMapper.java:58]: switch case is more compact here

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/PaymentMapper.java [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+        .amount(createPayment.getAmount())
+        .comment(createPayment.getComment())
+        .build();
+  }
+
+  public List<school.hei.haapi.model.Payment> toDomainPayment(
+      String feeId, List<CreatePayment> createPayment) {
+    Fee associatedFee = feeService.getById(feeId);
+    if (associatedFee == null) {
+      throw new BadRequestException("Fee.id=" + feeId + "is not found");
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/endpoint/rest/mapper/PaymentMapper.java:48]: NotFound

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/PaymentMapper.java [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+    }
+    List<school.hei.haapi.model.Payment> payments = new ArrayList<>();
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/endpoint/rest/mapper/PaymentMapper.java:50]: Functional

File: src/main/java/school/hei/haapi/model/Fee.java [hunk 19]
```diff
@@ -51,13 +57,17 @@ public class Fee implements Serializable {
...
+  @Transient
   private int remainingAmount;
 
   private String comment;
   @CreationTimestamp private Instant creationDatetime;
 
   private Instant dueDatetime;
 
+  @OneToMany(mappedBy = "fee")
+  private List<Payment> paymentList;
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/model/Fee.java:69]: payments

File: src/main/java/school/hei/haapi/service/FeeService.java [hunk 20]
```diff
@@ -3,57 +3,77 @@
...
+        amount += payment.getAmount();
+      }
     }
+    return fee.getTotalAmount() - amount;
   }
 
-  private void checkStudentFees(String studentId, List<Fee> fees) {
-    feeValidator.accept(fees);
-    fees.forEach(fee -> checkStudentFees(studentId, fee));
+  private Fee refreshFees(Fee fee) {
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/service/FeeService.java:68]: resetPaymentRelatedInfo

File: src/main/java/school/hei/haapi/service/FeeService.java [hunk 21]
```diff
@@ -3,57 +3,77 @@
...
   }
 
-  private void checkStudentFees(String studentId, Fee toCheck) {
-    if (!studentId.equals(toCheck.getStudentId())) {
-      throw new BadRequestException(
-          "Fee must be associated to student." + toCheck.getStudentId() + " instead of student."
-              + studentId);
+  private int computeRemainingAmount(Fee fee) {
+    List<Payment> payments = fee.getPaymentList();
+    int amount = 0;
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/service/FeeService.java:59]: use functional style, with sum() as final operation

File: src/main/java/school/hei/haapi/service/FeeService.java [hunk 22]
```diff
@@ -3,57 +3,77 @@
...
+      }
     }
+    return fee.getTotalAmount() - amount;
   }
 
-  private void checkStudentFees(String studentId, List<Fee> fees) {
-    feeValidator.accept(fees);
-    fees.forEach(fee -> checkStudentFees(studentId, fee));
+  private Fee refreshFees(Fee fee) {
+    fee.setRemainingAmount(computeRemainingAmount(fee));
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/service/FeeService.java:69]: build a new fee, and set the two fields in that builded fee. That way you don't mutate the original object

File: src/main/java/school/hei/haapi/service/PaymentService.java [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+
+  @Transactional
+  public List<Payment> saveAll(List<Payment> toCreate) {
+    checkStudentPayments(toCreate);
+    return paymentRepository.saveAll(toCreate);
+  }
+
+  private void checkStudentPayments(Payment toCheck) {
+    Fee associatedFee = toCheck.getFee();
+    if (associatedFee.getRemainingAmount() < toCheck.getAmount()) {
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/service/PaymentService.java:44]: No. This is a bug. Suppose your remaining amount is 10. But then you want to create 5 payments of 5 euros each. All 5 payments will be persisted as for each of them 5 < 10.

What you need to do is to sum the amount of all payments and to check remaining amount against it.

Put the scenario I described in our test suite.

File: src/main/resources/db/migration/V0_4__Create_payment_table.sql [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+    end
+$$;
+
+create table if not exists "payment"
+(
+    id                varchar
+        constraint payment_pk primary key default uuid_generate_v4(),
+    fee_id           varchar                  not null
+        constraint payment_fee_id_fk references "fee"(id),
+    type              payment_type                 not null,
```
Comment (hei-teacher) [inline src/main/resources/db/migration/V0_4__Create_payment_table.sql:16]: indentation

File: src/main/resources/db/testdata/V99_4__Test_create_payments.sql [hunk 25]
```diff
@@ -0,0 +1,10 @@
+insert into "payment"
+    (id, fee_id, type, comment,amount, creation_datetime)
+values
+    ('payment1_id','fee1_id', 'CASH', 'Comment', 2000, '2022-11-08T08:25:24.00Z'),
```
Comment (hei-teacher) [inline src/main/resources/db/testdata/V99_4__Test_create_payments.sql:4]: Use a richer set of data. Use all types. And sometimes set comment to `null`.

Also, add the MOBILE_MONEY type, which requires comment to be set when used (check is done on the Java side, not Postgres).

FIX type also requires comment to be set (to be checkd on Java side)

File: src/test/java/school/hei/haapi/integration/PaginationIT.java [hunk 26]
```diff
@@ -88,10 +105,48 @@ void student_pages_are_ordered_by_reference() throws ApiException {
     assertTrue(isBefore(page2.get(0).getRef(), page2.get(2).getRef()));
   }
 
+  @Test
+  void fees_page_are_ordered_by_due_datetime_desc() throws ApiException {
```
Comment (hei-teacher) [inline src/test/java/school/hei/haapi/integration/PaginationIT.java:109]: pages

File: src/test/java/school/hei/haapi/integration/PaginationIT.java [hunk 27]
```diff
@@ -88,10 +105,48 @@ void student_pages_are_ordered_by_reference() throws ApiException {
...
+
+    assertEquals(pageSize, page1.size());
+    assertEquals(1, page2.size());
+    assertEquals(0, page3.size());
+    assertTrue(isAfter(page1.get(0).getDueDatetime(), page1.get(1).getDueDatetime()));
+    assertTrue(isAfter(page1.get(1).getDueDatetime(), page2.get(0).getDueDatetime()));
+  }
+
+  @Test
+  void payments_page_are_ordered_by_due_datetime_desc() throws ApiException {
```
Comment (hei-teacher) [inline src/test/java/school/hei/haapi/integration/PaginationIT.java:126]: pages

File: src/test/java/school/hei/haapi/integration/PaginationIT.java [hunk 28]
```diff
@@ -88,10 +105,48 @@ void student_pages_are_ordered_by_reference() throws ApiException {
     assertTrue(isBefore(page2.get(0).getRef(), page2.get(2).getRef()));
   }
 
+  @Test
+  void fees_page_are_ordered_by_due_datetime_desc() throws ApiException {
+    ApiClient student1Client = anApiClient(STUDENT1_TOKEN);
+    PayingApi api = new PayingApi(student1Client);
+    int pageSize = 2;
+
+    List<Fee> page1 = api.getStudentFees(STUDENT1_ID, 1, pageSize);
```
Comment (hei-teacher) [inline src/test/java/school/hei/haapi/integration/PaginationIT.java:114]: Can you check whether we put ordered indexes (of the approriate orderding) on fee.due_datetime (DESC) and payment.creation_datetime (DESC or ASC, I don't remember)? https://www.postgresql.org/docs/current/indexes-ordering.html

When we now that we will ALWAYS (or very often) return a list ordered by some attribute, we add such ordered indexes. Warning: if the index order is DIFFERENT from the request submitted, then we may have dramatic result!

File: src/test/java/school/hei/haapi/integration/PaymentIT.java [hunk 29]
```diff
@@ -0,0 +1,207 @@
...
+    assertEquals(0, expected.size());
+    assertEquals(expected, actual);
+  }
+
+  @Test
+  void manager_write_with_some_bad_fields_ko() {
+    ApiClient manager1Client = anApiClient(MANAGER1_TOKEN);
+    PayingApi api = new PayingApi(manager1Client);
+    CreatePayment toCreate1 = creatablePayment1().amount(null);
+    CreatePayment toCreate2 = creatablePayment1().amount(-1);
```
Comment (hei-teacher) [inline src/test/java/school/hei/haapi/integration/PaymentIT.java:195]: Unless type is FIX! But just create a ticket for addressing FIX types for the moment

**Review by hei-teacher — CHANGES_REQUESTED** [review]

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeMapper.java (hunk 10 above)
Comment (hei-ryan) [reply src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeMapper.java:23]: In the getter for example ? (`getCreationDatetime()`)

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/PaymentMapper.java (hunk 14 above)
Comment (hei-ryan) [reply src/main/java/school/hei/haapi/endpoint/rest/mapper/PaymentMapper.java:21]: In Getter ?

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/PaymentMapper.java (hunk 16 above)
Comment (hei-ryan) [reply src/main/java/school/hei/haapi/endpoint/rest/mapper/PaymentMapper.java:58]: [https://stackoverflow.com/questions/3827393/java-switch-statement-constant-expression-required-but-it-is-constant](https://stackoverflow.com/questions/3827393/java-switch-statement-constant-expression-required-but-it-is-constant)

image

File: src/main/java/school/hei/haapi/service/PaymentService.java (hunk 23 above)
Comment (hei-ryan) [reply src/main/java/school/hei/haapi/service/PaymentService.java:44]: The check (5 <10) is only available for the 1st payment and during the 2nd payment, the fee's remaining amount is 5 not 10 anymore ... 

Concurrence ?

File: src/test/java/school/hei/haapi/integration/PaymentIT.java (hunk 29 above)
Comment (hei-teacher) [reply src/test/java/school/hei/haapi/integration/PaymentIT.java:195]: My above comment is bad. Indeed, amount can NEVER be negative. The FIX type is used to cancel fees that we don't want pay (eg: fees that we created falsly).

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeMapper.java (hunk 10 above)
Comment (hei-teacher) [reply src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeMapper.java:23]: Yes! That would be great 👍

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/PaymentMapper.java (hunk 16 above)
Comment (hei-teacher) [reply src/main/java/school/hei/haapi/endpoint/rest/mapper/PaymentMapper.java:58]: Directly switch over the enum value, not it's toString result.

File: src/main/java/school/hei/haapi/service/PaymentService.java (hunk 23 above)
Comment (hei-teacher) [reply src/main/java/school/hei/haapi/service/PaymentService.java:44]: Given fee1.remnaingAmount = 10

When pay(fee1, payments=[payment1=5, payment2=5, ..., payment5=5]

Then the whole payments list should be discarded. Which is currently not the case: your implementation will persist those 5 payments even though the will result in an overpayment of fee1.

File: src/test/java/school/hei/haapi/integration/PaginationIT.java (hunk 28 above)
Comment (hei-ryan) [reply src/test/java/school/hei/haapi/integration/PaginationIT.java:114]: Actually, we didn't index _yet_ the fee.due_datetime(DESC) and payment.creation_datetime(DESC). For our use case, the most recent (descending) order is the most efficient, and I don't think there is a case where we should change this order.

So I think we can put ordered (DESC) indexes for these fields.

File: src/test/java/school/hei/haapi/integration/PaymentIT.java (hunk 29 above)
Comment (hei-ryan) [reply src/test/java/school/hei/haapi/integration/PaymentIT.java:195]: For sure, but this feature is not still implemented so I've created a ticket anyway : [#60 ](https://github.com/hei-school/hei-admin-api/issues/60)

**Pull request merged**